Alwaywas return local url for users avatar

strk opened the pull request:
Avoids having to wait for DNS lookups when libravatar is activated
fixing #6046

strk commented:
this PR still has an issue with relative URLS returned instead of absolute ones

strk commented:
Now things are fixed. Note that this is also the start to eventually act as a proxy, rather than redirecting.

guillep2k commented:
Please do `make fmt` and push again.

strk commented:
I don't understand the unit test failure: https://drone.gitea.io/go-gitea/gitea/13134/1/11
Can anyone explain it to me ?

techknowlogick commented:
@strk 
Here is the specific test that failed:
```
--- FAIL: TestPushCommits_AvatarLink (0.01s)
action_test.go:175: 
    	Error Trace:	action_test.go:175
    	Error:      	Not equal: 
    	            	expected: "https://secure.gravatar.com/avatar/ab53a2911ddf9b4817ac01ddcd3d975f?d=identicon"
    	            	actual  : "/suburl//explore/avatars/user2/-1"
    	            	
    	            	Diff:
    	            	--- Expected
    	            	+++ Actual
    	            	@@ -1 +1 @@
    	            	-https://secure.gravatar.com/avatar/ab53a2911ddf9b4817ac01ddcd3d975f?d=identicon
    	            	+/suburl//explore/avatars/user2/-1
    	Test:       	TestPushCommits_AvatarLink
```

lafriks commented on routers/home.go:
```diff
@@ -380,6 +381,28 @@ func ExploreCode(ctx *context.Context) {
+
+	user, err := models.GetUserByName(userName)
+	if err != nil {
+		// if models.IsErrUserNotExist(err)
```
Commented out code should be removed

lafriks commented on routers/home.go:
```diff
@@ -380,6 +381,28 @@ func ExploreCode(ctx *context.Context) {
+		return
+	}
+
+	url := user.RealSizedAvatarLink(size)
```
No need for variable if it is used only once, just:
```suggestion
	ctx.Redirect(user.RealSizedAvatarLink(size))
```

strk commented:
Both suggestions implemented in
commit 2019e5005c9ed158e1bb47d3b98a347724364a3b

techknowlogick commented on routers/home.go:
```diff
@@ -380,6 +381,30 @@ func ExploreCode(ctx *context.Context) {
+		if models.IsErrUserNotExist(err) {
+			ctx.ServerError("Requested avatar for invalid user", err)
+		} else {
+			ctx.ServerError("Retriving user by name", err)
```
typo: `Retrieving`

strk commented:
Typo fixed

lafriks commented on routers/routes/routes.go:
```diff
@@ -267,6 +267,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 		m.Get("", func(ctx *context.Context) {
 			ctx.Redirect(setting.AppSubURL + "/explore/repos")
 		})
+		m.Get("/avatars/:username/:size", routers.ExploreAvatars)
```
imho it should be better placed under `/user` group:
https://github.com/go-gitea/gitea/blob/8246fe03968a6939159b65371570749f1698439c/routers/routes/routes.go#L403
```

so URL would look like `/user/avatar/:username/:size`

lafriks replied:
You have added to wrong "/user" group. See my previous link above for correct place

strk replied:
Now moved

lafriks commented:
Test case needs to be updated:
https://github.com/go-gitea/gitea/blob/b660a732ae283d863636ead9cc1a365ce1c0edc1/models/action_test.go#L176

This still will leave problem with git author emails that are not for any user on current instance but that should probably be solved in other PR

strk commented:
@lafriks may you help me find the obtained (vs. expected) value from https://drone.gitea.io/go-gitea/gitea/13174/1/11 ? I cannot find any match when I search for "gravatar" or "action_test" there

strk commented:
I've changed route as requested, and updated one of the tests, squash rebased to master and force-pushed. The case of external users' avatar is still to be decided.

lafriks commented on models/action_test.go:
```diff
@@ -173,7 +173,7 @@ func TestPushCommits_AvatarLink(t *testing.T) {
 
 	assert.Equal(t,
-		"https://secure.gravatar.com/avatar/ab53a2911ddf9b4817ac01ddcd3d975f?d=identicon",
+		"/user/avatar/user2/-1",
```
```suggestion
		"/suburl//user/avatar/user2/-1",
```

strk replied:
Fixed also removing the duplicated slash

strk commented:
For unknown users avatar, how about `/{org|user}/repo/commit/{hash}/avatar` ?

strk commented:
There's something wrong with the route, logs show that Gitea is trying to search a user called "user":

````
[Macaron] 2019-09-24 11:11:22: Started GET /user/strk/avatar/-1 for 127.0.0.1
2019/09/24 11:11:22 ...s/context/context.go:329:func1() [D] Session ID: 1168d515f9611a5d
2019/09/24 11:11:22 ...s/context/context.go:330:func1() [D] CSRF Token: Wl0kZl7NTbZRvAnr0k5UlzHV1uI6MTU2OTMxNTgxNzUyODIyMDAzMQ
2019/09/24 11:11:22 .../xorm/session_raw.go:76:queryRows() [I] [SQL] SELECT "id", "lower_name", "name", "full_name", "email", "keep_email_private", "email_notifications_preference", "passwd", "passwd_hash_algo", "must_change_password", "login_type", "login_source", "login_name", "type", "location", "website", "rands", "salt", "language", "description", "created_unix", "updated_unix", "last_login_unix", "last_repo_visibility", "max_repo_creation", "is_active", "is_admin", "allow_git_hook", "allow_import_local", "allow_create_organization", "prohibit_login", "avatar", "avatar_email", "use_custom_avatar", "num_followers", "num_following", "num_stars", "num_repos", "num_teams", "num_members", "visibility", "diff_view_style", "theme" FROM "user" WHERE "lower_name"=$1 LIMIT 1 []interface {}{"user"} - took: 1.82165ms
2019/09/24 11:11:22 routers/repo/http.go:97:HTTP() [E] GetUserByName: user does not exist [uid: 0, name: user, keyid: 0]
```

I'm thinking again we should have this under `/explore/avatar/` maybe with `/explore/avatar/user/:name` and `/explore/avatar/commit/org/repo/:hash`

strk commented:
Ok forget the report above, was something different. Current code works fine. Only left-out is the per-commit (unknown user) case. I think we don't want to stamp an email address on an URL do we ?

lafriks commented:
No we don't want email to be shown in url probably

lafriks commented:
I think that chase could be left for other pr

lunny commented:
It seems that this will only change a direct media request to a redirect media request. I don't think this will fix the problem in any case.

strk commented:
It fixes the problem of having to wait all DNS replies _before_ starting to receive the HTML page.
Instead the browser will _immediately_ get the HTML page and only wait for each of the images (some images will need to wait more, some less...)

strk commented:
Once this change is in, further improvement may imply local caching user's avatars (optional) for fixing the privacy issue raising some concerns around (ie: recognizing users by their email hash)

guillep2k commented:
> 
> 
> Once this change is in, further improvement may imply local caching user's avatars (optional) for fixing the privacy issue raising some concerns around (i.e.: recognizing users by their email hash)

_In theory_ the content provider should send an [`ETag`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) value. Caching explicitly should be unnecessary (but... you know, some providers might not be very collaborative).

strk commented:
CI-green and approved, can this be merged please ?
